### PR TITLE
rework the vtable implementation with IC in mind

### PR DESCRIPTION
**Pull request by ringabout**
follow up #21342

todo

- [x] register the type properly in sempass2
- [x] make it opt-in after some more testing

### CODE REVIEW

**Comment by ringabout** [comment]
Will investigate the failures this weekend

**Comment by Araq** [comment]
What's the status of this?

**Comment by ringabout** [comment]
It is ready for review, having reported it last Monday. I have made vtables opt-in.

File: compiler/cgen.nim [hunk 1]
```diff
@@ -2181,9 +2181,30 @@ proc updateCachedModule(m: BModule) =
   cf.flags = {CfileFlag.Cached}
   addFileToCompile(m.config, cf)
 
-proc finalCodegenActions*(graph: ModuleGraph; m: BModule; n: PNode): PNode =
+proc genVTable(m: BModule, seqs: seq[PSym]): Rope =
+  result = Rope"{"
+  for i in 0..<seqs.len-1:
+    result.add "(void *) " & seqs[i].loc.r & ", "
+  result.add "(void *) " & seqs[^1].loc.r
```
Comment (Araq) [inline compiler/cgen.nim]:
```suggestion
    if i > 0: result.add ", "
    result.add "(void *) " & seqs[i].loc.r
```

File: compiler/cgen.nim [hunk 2]
```diff
@@ -2181,9 +2181,30 @@ proc updateCachedModule(m: BModule) =
   cf.flags = {CfileFlag.Cached}
   addFileToCompile(m.config, cf)
 
-proc finalCodegenActions*(graph: ModuleGraph; m: BModule; n: PNode): PNode =
+proc genVTable(m: BModule, seqs: seq[PSym]): Rope =
+  result = Rope"{"
+  for i in 0..<seqs.len-1:
```
Comment (Araq) [inline compiler/cgen.nim]:
```suggestion
  for i in 0..<seqs.len:
```

**Comment by ringabout** [comment]
succeeded by https://github.com/nim-lang/Nim/pull/22991

**Pull request closed without merging**